### ISSUE TITLE
Not able to handle when server ends the connection

### ISSUE THREAD

**Issue by woopstar** [tgriesser/knex#1198]
I'm seeing this problem because there is a connection timeout for 600 seconds from the MySQL server. Then a connection is closed from the server instead of the client.

When Knex then tried to execute a query, the following error comes:

```
Unhandled rejection Error: BEGIN; - This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (net.js:297:12)
    at Connection.write (/node_modules/mysql2/lib/connection.js:136:15)
    at Connection.writePacket (/node_modules/mysql2/lib/connection.js:150:10)
    at Query.start (/node_modules/mysql2/lib/commands/query.js:39:14)
    at Query.Command.execute (/node_modules/mysql2/lib/commands/command.js:34:20)
    at Connection.handlePacket (/node_modules/mysql2/lib/connection.js:310:28)
    at Connection.addCommand (/node_modules/mysql2/lib/connection.js:326:10)
    at Connection.query (/node_modules/mysql2/lib/connection.js:397:15)
    at /node_modules/knex/lib/dialects/mysql/index.js:92:18
    at tryCatcher (/node_modules/knex/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._resolveFromResolver (/node_modules/knex/node_modules/bluebird/js/main/promise.js:480:31)
    at new Promise (/node_modules/knex/node_modules/bluebird/js/main/promise.js:70:37)
    at Client._query (/node_modules/knex/lib/dialects/mysql/index.js:88:12)
    at Client.query (/node_modules/knex/lib/client.js:127:24)
    at /node_modules/knex/lib/transaction.js:219:21
    at tryCatcher (/node_modules/knex/node_modules/bluebird/js/main/util.js:26:23)
```

**Comment by wubzz** [comment]
Vague interpretation of what you mean by "Not able to handle" I'm guessing you say that because the error was thrown rather than handled? (Missing `.catch` block?)

Or are you saying knex should attempt to open up a new connection when this occurs?

**Comment by woopstar** [comment]
I believe Knex should attempt to open a connection when this occours. If a pool object exists, but the sql server has closed the connection and you're trying to send a query. In stead of failing, it should first attempt to reconnect and if not able to do so, then fail?

**Comment by elhigu** [comment]
@woopstar I agree knex should be able to handle that case and just throw away the connection and try to get new one if DB side had closed it. Then if reconnection doesn't work pool can throw connection timeout exception...

**Comment by woopstar** [comment]
@elhigu I agree. The use of pools is pointless in my opinion if knex does not use a reconnect/new pool logic when executing a query. The only reason to throw an exception would simply be if knex is not possible to reach the sql server. There is tons of cases where the pool connection can be closed in between queries.
We often end connections after 600 seconds to prevent transaction deadlocking etc.

**Comment by wubzz** [comment]
With the new `acquireConnectionTimeout` option as elhigu mentioned this makes a lot of sense. I'll see if I can look into this, unless someone beats me to it.

**Comment by elhigu** [comment]
@wubzz I was talking about this issue with a friend in work and sounds like this feature actually cannot work at all by just retrying failed request. Retry would work only for single request, e.g. for transactions it would be pretty much impossible to implement, since after getting new connection all the requests in the transaction should be retried.

However Pool2 should be setup to use it's ping feature to make sure that connections, which are inside the pool are actually alive. 

I'm closing the #1200 pull request, but still we could make pull request, which makes `knex` to use `pool2`'s ping method configuration to make sure that connections in pool are alive.

**Comment by wubzz** [comment]
@elhigu That makes sense, didn't think about transactions at all.. Will have a look at pool2 later.

**Comment by wubzz** [comment]
@elhigu Playing around with `pool2`, the ping function was by far a better way of doing what I tried to achieve in that PR:

```javascript
pool: {
		ping: function(connection, callback) {
			connection.query({text: 'SELECT 1 = 1'}, [], callback);
		},
		min: 1,
		max: 2
	}
```

Creates a new connection when needed, as expected.

However, the issue of transactions remains. Given the following example:

```javascript
knex.transaction(function(tr) {
	tr.raw('SELECT version()')
	.then(function() {
			tr.raw('SELECT version()')
			.then(tr.commit)
			.catch(tr.rollback);
	})

}).then(function() {
	console.log('Transaction Comitted');
}).catch(function(error) {
	console.log('Transaction Rollback', error);
});
```

In this code, the `ping` function isn't called _(this is correct)_, instead an error is thrown `ECONNRESET`, and the `rollback` function is called. In turn, this will attempt to run a `ROLLBACK;` query using the same connection that was just severed, which obviously won't work. Instead, it just hangs with what seems to be without a timeout.

One wonky solution is to just throw the error down the chain to the transaction's `reject` handler when the `ECONNRESET` error is thrown rather than trying to issue `ROLLBACK;` query, but perhaps there is a better way of handling this? I'm not sure..

**Comment by elhigu** [comment]
Uh... so if connection is closed running `trx.rollback`does not fail, but just hang... Does commit / other queries work the same way (just hang) that they just hang on if they are ran on closed connection? I don't understand why `rollback` would do anything different. I haven't looked into transaction code for a while so I cannot ay what is going on with that. 

To me it would sound ok if `knex` just ignores any error on `rollback` and keeps on running since if rollback fails, there is not much what one can do. Then pool2 can just dispose broken connection with its ping functionality. @rhys-vdw @tgriesser any insight on this one?

**Comment by wubzz** [comment]
Related issues: #1072 #892

**Comment by tmeinders** [comment]
Thanks @wubzz for the info on #1072.

I'm trying to see how the pinging approach could bring the stack (from knex down to the mysql client) to a level of reliability that is practical. Conceptually, I see that pinging will accomplish two things: 
1. keep alive connections
2. discovery of dead connections

Most of the time, pinging will achieve the first goal: keeping the connections alive. This change alone will help prevent a lot of the issues I'm sure we are all running into.

However, connections could be dropped for other reasons and using pinging to find dead connections is a lot like using polling to do a health check. We are basically hoping for a favorable outcome from a race condition. We are depending on chance that the ping will expose the bad connection before the core application tries to use it.

**Comment by woopstar** [comment]
We've implemented the ping process in our environment. The above example is a bit bogus, but the following works:

```
pool: {
        ping: function(connection, callback) {
            connection.query({sql: 'SELECT 1 = 1'}, [], callback);
        },
        pingTimeout: 3*1000,
        min: 1,
        max: 10
    }
```

It seems to have solved our problems for now.

**Comment by wubzz** [comment]
@tmeinders Unless the connection is dropped right after the `ping` function has verified that the connection is still valid, then what you described should not be a problem. The `ping` function, as I understand, is not in any way "polling" the database server, but is instead called *before* a query is run using the connection. This means dead connections *should* always be discovered before the connection is used to run a query, or as you described it, before the core application tries to use it.

Currently I believe this is the closest to a good solution we have, disregarding the issue of transactions, which quite frankly, one can't really do anything about to recover the state anyway. 

@elhigu @rhys-vdw There's around 5 issues in this repository that are relevant to this. My proposal is this:
- Add default `ping` method for all dialects as to avoid these issues as much as possible. It's a part of Pool2 for a reason. *(See #922 )*
- Add a reasonable timeout to transactions `rollback;` query in case its being called on a dead connection, and if the timeout exceeds, simply ignore the error in order to release the connection to the pool. *(Connection broken, and transaction not commited so there's no harm, and can't recover the state)*

If a user does not want knex to issue a `ping` query such as `SELECT 1` to verify the connection, then they can supply their own `ping` function in the config following Pool2's documentation.

Does this sound good to you?

**Comment by elhigu** [comment]
@wubzz sounds good!

About rollback timeout, if timeout triggers could we make sure that the connection will be discarded from the pool?

**Comment by wubzz** [comment]
@elhigu Yes, that should not be an issue. I'm thinking something like

```javascript
.rollback()
.timeout(some ms)
.catch(Promise.TimeoutError, (err) => {
//Release connection to pool
//Resolve rollback promise as if nothing happened.
})
.catch((err) => {
//Still throws error as usual when the explicit Timeout isn't triggered.
});
```

**Comment by woopstar** [comment]
So. I guess it does not work yet. I've implemented the following code:

```
var knex = require('knex')({
    client: 'mysql2',
    connection: {
        host: mysqlConfig.master.host,
        user: mysqlConfig.master.username,
        password: mysqlConfig.master.password,
        database: mysqlConfig.master.database
    },
    pool: {
        ping: function(connection, callback) {
            connection.query({sql: 'SELECT 1 = 1'}, [], callback);
        },
        pingTimeout: 3*1000,
        min: 1,
        max: 10
    }
});
```

But it does still not work. The timeout on our database server is 600 sec. And when the connection is timed out, the following still occur:

```
Error: This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (net.js:268:12)
    at Connection.write (/data/nodejs/145ac91/node_modules/mysql2/lib/connection.js:136:15)
    at Connection.writePacket (/data/nodejs/145ac91/node_modules/mysql2/lib/connection.js:150:10)
    at Query.start (/data/nodejs/145ac91/node_modules/mysql2/lib/commands/query.js:39:14)
    at Query.Command.execute (/data/nodejs/145ac91/node_modules/mysql2/lib/commands/command.js:34:20)
    at Connection.handlePacket (/data/nodejs/145ac91/node_modules/mysql2/lib/connection.js:310:28)
    at Connection.addCommand (/data/nodejs/145ac91/node_modules/mysql2/lib/connection.js:326:10)
    at Connection.query (/data/nodejs/145ac91/node_modules/mysql2/lib/connection.js:397:15)
    at Pool.require.pool.ping [as _ping] (/data/nodejs/145ac91/lib/db.js:22:24)
    at Pool._maybeAllocateResource (/data/nodejs/145ac91/node_modules/pool2/lib/pool.js:389:18)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
```

From the following code:

```
knex
        .select('id', 'number', 'country', 'full_number', 'closed')
        .from('dids')
        .where('tenants_id', req.tenant.id)
        .then(function (rows) {
            if (rows.length > 0) {
                res.sendData(rows);
            } else {
                res.send(404);
            }
            next();
        })
        .catch(function (error) {
            res.send(500, error);
            next();
        });
```

**Comment by woopstar** [comment]
Tried this, but didn't work either. Though if I did a dirty hack and pack the connection query into a try catch and the call the callback on errors. As ping disposes the object on error. But it seems it doesn't create a new one to send the query through.

```
pool: {
        ping: function(connection, callback) {
            try
            {
                connection.query({sql: 'SELECT 1 = 1'}, [], callback);
            }
            catch (e)
            {
                callback(e);
            }
        },
        pingTimeout: 3*1000,
        min: 1,
        max: 10
    }
```

**Comment by woopstar** [comment]
Also tried this, and not working:

```
pool: {
        ping: function ping(connection, callback) {
            try {
                connection.ping(callback);
            } catch (err) {
                process.nextTick(callback.bind(null, err));
            }
        },
        min: 1,
        max: 2
    }
```

**Comment by wubzz** [comment]
@woopstar I was finally able to reproduce this error setting `connect_timeout`, `wait_timeout` and `interactive_timeout` to `1` on my mysql server.

`COMMIT; - This socket has been ended by the other party`

I can confirm the `ping` solution works when **not** using transactions.

Transactions are bound to their single connection and therefor `ping` is not being called on these. Because of this, it eventually throws that error. I don't see how knex can possibly recover from this and "rerun" the entire transaction using a new connection.

From your original post, the error was
`Unhandled rejection Error: BEGIN; - This socket has been ended by the other party`
indicating that you are using a transaction _(BEGIN;)_, however in your code example above I do not see a transaction being used. Can you post your code that causes this issue while inside a transaction?

If you have a timeout of 600 seconds (10 minutes) on your mysql server, there should from my understanding be no way for a transaction to throw this error so long as the transaction is either commited or rolled back within those 10 minutes.

**Comment by woopstar** [comment]
Sorry for late reply.

We use transactions for all queries that alter data on the server - so yes, we are using transactions and this is causing the problem.

The problem does not rely on that the transaction takes 600 seconds, but that the connection from the server is closed after 600 seconds. When the transaction is starting, it tries to use a pool object that is actually closed and this causes the issue.
The only solution we found yet, is to query ever 500 second to keep the connection open and used.

When a transaction or a query starts / is fired, the pool object must be validated and checked if it's actually still connected. It thinks it is, because it dosen't see that the server closed the connection.

Makes sense?

**Comment by wubzz** [comment]
@woopstar Gotcha. The only thing that does **not** make sense is how the Pooling fix above does not fix your issue.

The `SELECT 1` check should occur *before* the connection calls for `BEGIN;` and with that it should realize the connection is closed since the query fails, and instead return a fresh connection..

I'm not able to reproduce what you described even with a 1 second timeout so I'm a bit stumped..

**Comment by woopstar** [comment]
We tried with all of the above, and all of them gave us the error about the connection to the server was closed. I agree with you, that the ping method should verify the connection before the query is actually run against it, but it seems not to be the case. Or maybe it does, but does not open up a new connection and try that one.

**Comment by woopstar** [comment]
This is our currently code, which seems to work the most. Though we got the following error today:

`Mar 29 06:37:02 odn1-voip-api-v2-01 API: {"name":"api","hostname":"odn1-voip-api-v2-01","pid":26137,"level":50,"msg":"Knex heartbeat error { [Error: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?]\n  timeoutStack: 'TimeoutError: operation timed out\\n    at afterTimeout (/data/nodejs/4568574/node_modules/bluebird/js/main/timers.js:16:15)\\n    at timeoutTimeout (/data/nodejs/4568574/node_modules/bluebird/js/main/timers.js:59:9)\\n    at Timer.listOnTimeout (timers.js:92:15)',\n  sql: 'SELECT 1=1',\n  bindings: undefined }","time":"2016-03-29T04:37:02.089Z","v":0}`

-

```
'use strict';

var config = require('config');
var log = require('./logger');
var domain = require('domain').create();

var mysqlConfig = config.mysql;

if (!mysqlConfig) {
    console.error('No MySQL config found, exiting');
    process.exit(1);
}

var heartBeatInterval = null;

function createKnex() {
    var rVal = require('knex')({
        client: 'mysql2',
        connection: {
            host: mysqlConfig.master.host,
            user: mysqlConfig.master.username,
            password: mysqlConfig.master.password,
            database: mysqlConfig.master.database
        },
        pool: {
            ping: function ping(connection, callback) {
                try {
                    connection.ping(callback);
                } catch (err) {
                    log.error('In catch', err);
                    process.nextTick(callback.bind(null, err));
                }
            },
            min: 1,
            max: 2
        }
    });
    
    if (heartBeatInterval) {
        clearInterval(heartBeatInterval)
        heartBeatInterval = null;
    }
    
    heartBeatInterval = setInterval(function () {
        rVal.raw('SELECT 1=1')
        .then(function () {
          //  log.info('heartbeat sent');
        })
        .catch(function (err) {
            log.error('Knex heartbeat error', err);
        })
    }, 10000);
    
    return rVal;
}

domain.on('error', function (err) {
    log.error('knex error', err);
})

domain.run(function () {
    var knex = createKnex();
    module.exports = knex;
});
```


Why the SELECT 1=1 would give this error seems very strange to me.

**Comment by wubzz** [comment]
Ok so transaction much like anything else calls for `client.acquireConnection` which in turn calls for `client.pool.acquire` which finally calls the dialect-specific `acquireRawConnection`.

Now I don't know much about mysql but this diff between mysql and mysql2 strikes me as pretty odd.. Could this have anything to do with the errors you are seeing?

https://github.com/tgriesser/knex/blob/master/src/dialects/mysql/index.js#L52
https://github.com/tgriesser/knex/blob/master/src/dialects/mysql2/index.js#L64

mysql has listeners for `error` and `end` to dispose of the connection. mysql2 does not have this.

```javascript
connection.on('error', connectionErrorHandler.bind(null, client, connection))
connection.on('end', connectionErrorHandler.bind(null, client, connection))
```

According to [mysql2](https://www.npmjs.com/package/mysql2) the `connection` *does* in fact support the `end` event.

@elhigu Any idea about this?

**Comment by woopstar** [comment]
You think i should try the mysql lib instead of the mysql2 lib?

**Comment by wubzz** [comment]
@woopstar It's worth a shot to see if it solves your issue. I'll look into the diff between these two later today.

**Comment by wubzz** [comment]
@woopstar Both PRs are now merged in prepration for 0.11. Did you find any success in moving to mysql instead of mysql2?

**Comment by elhigu** [comment]
Closing in favor of #1833 (+ pool has changed after this issue)
